File: windows/CellBlock.Shared/Models/ApiModels.cs
using System.Text.Json.Serialization;

namespace CellBlock.Shared.Models;

/// <summary>
/// Standard API response wrapper
/// </summary>
public class ApiResponse<T>
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("data")]
    public T? Data { get; set; }

    [JsonPropertyName("error")]
    public string? Error { get; set; }

    [JsonPropertyName("message")]
    public string? Message { get; set; }
}

/// <summary>
/// Device platform types
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum DevicePlatform
{
    Ios,
    Windows,
    Android,
    Macos
}

/// <summary>
/// Device registration/status model
/// </summary>
public class Device
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("userId")]
    public string UserId { get; set; } = string.Empty;

    [JsonPropertyName("deviceFingerprint")]
    public string DeviceFingerprint { get; set; } = string.Empty;

    [JsonPropertyName("platform")]
    public DevicePlatform Platform { get; set; }

    [JsonPropertyName("deviceName")]
    public string? DeviceName { get; set; }

    [JsonPropertyName("osVersion")]
    public string? OsVersion { get; set; }

    [JsonPropertyName("appVersion")]
    public string? AppVersion { get; set; }

    [JsonPropertyName("firstSeen")]
    public DateTime FirstSeen { get; set; }

    [JsonPropertyName("lastSeen")]
    public DateTime LastSeen { get; set; }

    [JsonPropertyName("isActive")]
    public bool IsActive { get; set; }
}

/// <summary>
/// Whitelist item model
/// </summary>
public class WhitelistItem
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("userId")]
    public string UserId { get; set; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("windowsDomain")]
    public string? WindowsDomain { get; set; }

    [JsonPropertyName("category")]
    public string Category { get; set; } = string.Empty;

    [JsonPropertyName("isEnabled")]
    public bool IsEnabled { get; set; }

    [JsonPropertyName("createdAt")]
    public DateTime CreatedAt { get; set; }
}

/// <summary>
/// Time budget configuration
/// </summary>
public class TimeBudget
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("userId")]
    public string UserId { get; set; } = string.Empty;

    [JsonPropertyName("dayOfWeek")]
    public int? DayOfWeek { get; set; }

    [JsonPropertyName("isWeekend")]
    public bool? IsWeekend { get; set; }

    [JsonPropertyName("minutesAllowed")]
    public int MinutesAllowed { get; set; }

    [JsonPropertyName("weeklyMaxMinutes")]
    public int? WeeklyMaxMinutes { get; set; }
}

/// <summary>
/// User authentication credentials
/// </summary>
public class LoginRequest
{
    [JsonPropertyName("email")]
    public string Email { get; set; } = string.Empty;

    [JsonPropertyName("password")]
    public string Password { get; set; } = string.Empty;
}

/// <summary>
/// Authentication response with tokens
/// </summary>
public class AuthResponse
{
    [JsonPropertyName("accessToken")]
    public string AccessToken { get; set; } = string.Empty;

    [JsonPropertyName("refreshToken")]
    public string RefreshToken { get; set; } = string.Empty;

    [JsonPropertyName("user")]
    public User User { get; set; } = new();
}

/// <summary>
/// User profile model
/// </summary>
public class User
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("email")]
    public string Email { get; set; } = string.Empty;

    [JsonPropertyName("displayName")]
    public string? DisplayName { get; set; }

    [JsonPropertyName("timezone")]
    public string Timezone { get; set; } = "UTC";

    [JsonPropertyName("isEmailVerified")]
    public bool IsEmailVerified { get; set; }
}

File: ios/CellBlock/Sources/Models/APIModels.swift
//
//  APIModels.swift
//  CellBlock
//
//  API Request/Response models
//

import Foundation

// MARK: - API Response
struct APIResponse<T: Codable>: Codable {
    let success: Bool
    let data: T?
    let error: String?
    let message: String?
}

// MARK: - Authentication
struct LoginRequest: Codable {
    let email: String
    let password: String
}

struct SignupRequest: Codable {
    let email: String
    let password: String
    let displayName: String?
}

struct AuthResponse: Codable {
    let accessToken: String
    let refreshToken: String
    let user: User
}

// MARK: - Device Registration
struct RegisterDeviceRequest: Codable {
    let deviceFingerprint: String
    let platform: DevicePlatform
    let deviceName: String?
    let osVersion: String?
    let appVersion: String?
}

// MARK: - Heartbeat
struct HeartbeatRequest: Codable {
    let deviceId: String
    let isWhitelistedApp: Bool
    let timestamp: Int // Unix timestamp
}

struct HeartbeatResponse: Codable {
    let remainingSeconds: Int
    let weeklyRemaining: Int
    let isLocked: Bool
}

// MARK: - Whitelist
struct AddWhitelistItemRequest: Codable {
    let name: String
    let iosBundleId: String?
    let windowsDomain: String?
    let androidPackageName: String?
    let comment: String?
}

struct RemoveWhitelistItemRequest: Codable {
    let itemId: String
    let comment: String?
}

// MARK: - Time Budget
struct UpdateTimeBudgetRequest: Codable {
    let mode: String // "per_day" or "weekday_weekend"
    let perDayLimits: [DayLimit]?
    let weekdayMinutes: Int?
    let weekendMinutes: Int?
    let weeklyMaxMinutes: Int
    let comment: String?

    struct DayLimit: Codable {
        let dayOfWeek: Int // 0-6
        let minutesAllowed: Int
    }
}

// MARK: - Warden
struct InviteWardenRequest: Codable {
    let email: String
    let isPrimary: Bool
}

struct ApproveRequestRequest: Codable {
    let requestId: String
    let approved: Bool
    let comment: String?
}

struct GrantParoleRequest: Codable {
    let inmateId: String
    let type: String // "minutes" or "until"
    let value: ParoleValue
    let reason: String?

    enum ParoleValue: Codable {
        case minutes(Int)
        case datetime(String)

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            if let int = try? container.decode(Int.self) {
                self = .minutes(int)
            } else if let string = try? container.decode(String.self) {
                self = .datetime(string)
            } else {
                throw DecodingError.typeMismatch(
                    ParoleValue.self,
                    DecodingError.Context(
                        codingPath: decoder.codingPath,
                        debugDescription: "Expected Int or String"
                    )
                )
            }
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            switch self {
            case .minutes(let int):
                try container.encode(int)
            case .datetime(let string):
                try container.encode(string)
            }
        }
    }
}

struct TriggerLockdownRequest: Codable {
    let inmateId: String
    let gracePeriodMinutes: Int?
    let reason: String?
}

// MARK: - Pagination
struct PaginatedResponse<T: Codable>: Codable {
    let data: [T]
    let total: Int
    let page: Int
    let limit: Int
    let totalPages: Int
}
